README.md
# Koop-Auth-Direct-File
## A authentication module for implementing direct authentication from client to Koop server with a file-based user-store

[![npm][npm-image]][npm-url]  

## Usage

To use this plugin, you need to create a JSON file store.  This should be an array of objects with properties `username` and `password`.  See [user-store.json.example](user-store.json.example) as an example. Note, that this is a terrible way to store credentials!  This plugin is really just meant as an example of how to design a Koop authorization plugin and not meant for any production environment.

This module exports an initialization function that must be supplied with a secret-key and the path to the file that is the user-store.  It will return a plugin object that can be registered with Koop.

```javascript
    const authPlugin = require('@koopjs/auth-direct-file')('your-secret-key', `path/to/user-store.json`, { tokenExpirationMinutes: 300 })
    
    koop.register(auth)
```

### Registration order

For an auth-plugin to secure the data served by a provider, it must be registered _before_ the provider.  Any providers registered _after_ the auth-plugin will not have the plugin's authorization code applied.

```javascript
    const Koop = require('@koopjs/koop-core');
    const providerOne = require('koop-provider-one');
    const providerTwo = require('koop-provider-two');
    const authPlugin = require('@koopjs/auth-direct-file')('your-secret-key', `path/to/user-store.json`);

    const koop = new Koop();

    // providerOne, registered before auth-plugin, will be secured by auth-plugin methods
    koop.register(providerOne);

    koop.register(authPlugin);

    // providerTwo, registered after auth-plugin, will NOT be secured
    koop.register(providerTwo);

    const provider = require('./')
    koop.register(provider)
``` 

The underlying reason that order matter is because an auth-plugin add its `authorize` and `authenticate` functions to all _registered_ providers. (Note that providers may implement their own `authorize` and `authenticate` prototype methods, and if they do these will be used preferentially.) Output services will leverage these functions to secure the service endpoints and properly route requests to authenticate.


## API

### Module initialization function 

`(secret, userStoreFilePath, options) ⇒ Object`
* configure the authentication module with secret use for token encoding/decoding

| Param | Type | Description |
| --- | --- | --- |
| secret | <code>string</code> | secret for encoding/decoding tokens |
| userStoreFilePath | <code>string</code> | path to the JSON file containing the array of username/password objects |
| options | <code>object</code> | options object |
| options.tokenExpirationMinutes | <code>integer</code> | minutes until token expires (default 60) |


## Notes on use with ArcGIS Online and ArcGIS Portal  
This authorization plugin has been tested with ArcGIS Online and ArcGIS Enterprise. Note that these ArcGIS products appear to block or be impaired by SSH-tunneling tools like `ngrok`, so if you are trying to test a local Koop instance against Online or Enterprise, it may not work as otherwise expected.  Generally, you'll have to do a "real" deployment to actually test authorization against these products.

For enterprise versions of Portal earlier than 10.6, you may need to [import the root of your certificate into Portal's trust store](http://enterprise.arcgis.com/en/portal/10.5/administer/linux/import-a-certificate-into-the-portal.htm). We have observed the inability to store credentials for a secured Koop service on Portal instances that have not yet imported the root SSL certificate (of the Koop instance) into the trust-store.

[npm-image]: https://img.shields.io/npm/v/@koopjs/auth-direct-file.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@koopjs/auth-direct-file
[travis-image]: https://travis-ci.org/koopjs/koop-auth-direct-file.svg?style=flat-square


## General notes on the authentication pattern used here

The authentication module implemented here uses a *direct authentication* pattern; it receives user credentials (username/password) from a client and authenticates those credentials against an identity/user-store. Requests with valid credentials are issued an access-token (a string of encoded-data); The access token is encoded with the use of a secret known only to the Koop server. The access-token expires and becomes invalid after a certain period (default of 60 minutes).

![get access token](https://gist.githubusercontent.com/rgwozdz/e44f3686abe40360532fbcc6dccf225d/raw/9768df32fc62e99ce7383c124cab8efdf45b1e18/koop-direct-auth-access-token.png)

In addition to directly accepting username and password, the `authenticate` method will also accept a valid token.  If receivied, a new token (with new expiry time) is issued and returned.

The issued access-token should be attached to all subsequent service requests by the client. When the server receives a request, it will check for the presence of an access-token and reject any requests that are missing such token. If the token is present, the server attempts to decode it with its stored secret. Failure to decode results in a request rejection. Once decoded, the server checks the token's expiration-date and rejects any token with a date that is out of range. If the token is not expired, the request for the desired resource proceeds.

![enter image description here](https://gist.githubusercontent.com/rgwozdz/e44f3686abe40360532fbcc6dccf225d/raw/9768df32fc62e99ce7383c124cab8efdf45b1e18/koop-direct-auth-resources.png)



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 3.0.0
### Changed
* eliminate authenticationSpecification method

### Added
* allow authenticate method to "refresh" a token
* allow token to arrive on header, query, body

### Fixed
* strip 'Bearer ' string from any authorization header

## 2.0.3
### Fixed
* username and password can arrive on query or body

## 2.0.2
* patch - invalidate with user not found

## 2.0.1
* migrate to jest

## 2.0.0
### Changed
* `authorize` function argument is now the full Express request object
* Authorization token may be delivered by query parameter `token` or the `authorization` header
* `authenticate` function argument is now the full Express request object
* `authenticate` function rejects if username or password are missing
* Update koop-output-geoservices peer dependency to 2.0.0

## [1.2.0] - 2018-06-08
### Added
* replaced `getAuthenticationSpecification` with `authenticationSpecification`

## [1.1.1] - 2018-05-31
### Fixed
* Include peer-dependencies koop and and koop-output-geoservices

## [1.1.0] - 2018-05-29
### Added
* new option `useHttp`. Must be a boolean. This will get added verbatim to the result of authenticationSpecification function

## [1.0.0] - 2018-05-22
### Added
* Initial release of an authentication plugin for Koop that leverages a file-based user-store.

[2.0.0]: https://github.com/koopjs/koop-auth-direct-file/compare/v1.2.0...v2.0.0
[1.2.0]: https://github.com/koopjs/koop-auth-direct-file/compare/v1.1.0...v1.2.0
[1.1.1]: https://github.com/koopjs/koop-auth-direct-file/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/koopjs/koop-auth-direct-file/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/koopjs/koop-auth-direct-file/releases/tag/v1.0.0


test/auth.js
const test = require('tape')
const path = require('path')
const jwt = require('jsonwebtoken')
const providerMock = {
  name: 'test-provider',
  Model: function () {}
}
const secret = 'secret'
const auth = require('../src')(secret, path.join(__dirname, '/fixtures/user-store.json'))

test('authorize success with token as query parameter', async function (t) {
  t.plan(1)
  // Mock token
  const token = jwt.sign({exp: Math.floor(Date.now() / 1000) + 120, iss: providerMock.name, sub: 'username'}, secret)
  let decoded = await auth.authorize({ query: { token } })
  t.equals(decoded.iss, providerMock.name)
})

test('authorize success with token as authorization header', async function (t) {
  t.plan(1)
  // Mock token
  const token = jwt.sign({exp: Math.floor(Date.now() / 1000) + 120, iss: providerMock.name, sub: 'username'}, secret)
  let decoded = await auth.authorize({ headers: { authorization: token } })
  t.equals(decoded.iss, providerMock.name)
})

test('authorize failure - no token', async function (t) {
  t.plan(1)
  try {
    await auth.authorize(undefined)
  } catch (err) {
    t.equals(err.code, 401)
  }
})

test('authorize failure - expired token', async function (t) {
  t.plan(1)
  // Mock token
  const token = jwt.sign({exp: Math.floor(Date.now() / 1000) - 120, iss: providerMock.name, sub: 'username'}, secret)
  try {
    await auth.authorize({ query: { token } })
  } catch (err) {
    t.equals(err.code, 401)
  }
})

test('authenticate success', async function (t) {
  t.plan(2)
  let result = await auth.authenticate({ query: { username: 'jerry', password: 'garcia' } })
  t.equals(typeof result.token, 'string')
  t.equals(typeof result.expires, 'number')
})

test('authenticate failure - missing username', async function (t) {
  t.plan(2)
  try {
    await auth.authenticate({ query: { password: 'garcia' } })
  } catch (err) {
    t.equals(err.code, 401)
    t.equals(err.message, 'Invalid credentials.')
  }
})

test('authenticate failure - missing password', async function (t) {
  t.plan(2)
  try {
    await auth.authenticate({ query: { username: 'jerry' } })
  } catch (err) {
    t.equals(err.code, 401)
    t.equals(err.message, 'Invalid credentials.')
  }
})

test('authenticate failure', async function (t) {
  t.plan(2)
  try {
    await auth.authenticate({ query: { username: 'lou', password: 'reed' } })
  } catch (err) {
    t.equals(err.code, 401)
    t.equals(err.message, 'Invalid credentials.')
  }
})

test('authenticationSpecifiction', t => {
  t.plan(1)
  let result = auth.authenticationSpecification()
  t.equals(result.useHttp, false)
})

test('authSpecOptions - useHttp: true', t => {
  t.plan(1)
  let optionAuth = require('../src')(secret, path.join(__dirname, '/fixtures/user-store.json'), {useHttp: true})
  let result = optionAuth.authenticationSpecification()
  t.equals(result.useHttp, true)
})

test('authSpecOptions - useHttp: false', t => {
  t.plan(1)
  let optionAuth = require('../src')(secret, path.join(__dirname, '/fixtures/user-store.json'), {useHttp: false})
  let result = optionAuth.authenticationSpecification()
  t.equals(result.useHttp, false)
})

test('tokenExpirationMinutes - invalid "tokenExpirationMinutes" setting', t => {
  t.plan(1)
  t.throws(function () {
    require('../src')(secret, path.join(__dirname, '/fixtures/user-store.json'), {tokenExpirationMinutes: -1})
  }, /"tokenExpirationMinutes" must be an integer >= 5/)
})

test('tokenExpirationMinutes - invalid "useHttp" setting', t => {
  t.plan(1)
  t.throws(function () {
    require('../src')(secret, path.join(__dirname, '/fixtures/user-store.json'), {useHttp: 'string-value'})
  }, /"useHttp" must be a boolean/)
})


src/index.spec.js
const helpers = require('./validate-credentials');
const path = require('path');
const jwt = require('jsonwebtoken');
const secret = 'secret';

jest.spyOn(helpers, 'validateCredentials');
jest.spyOn(jwt, 'sign');
jest.spyOn(jwt, 'verify');

describe('Auth Plugin', () => {
  describe('initialize plugin', () => {
    test('should return expected registration object', () => {
      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );
      expect(authPlugin.type).toEqual('auth');
      expect(typeof authPlugin.authenticate).toBe('function');
    });

    test('should fail to initialize due to file not found', () => {
      try {
        require('./index.js')(
          secret,
          path.join(__dirname, '../test/fixtures/userz-store.json'),
        );
        fail('should have thrown');
      } catch (error) {
        expect(error.message).toMatch(
          /userz-store.json not found$/,
        );
      }
    });
  });

  describe('authenticate', () => {
    test('should fail to validate', async () => {
      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );

      try {
        await authPlugin.authenticate({
          query: { username: 'foo', password: 'bar' },
        });
        fail('should have thrown');
      } catch (error) {
        expect(error.message).toEqual('Invalid credentials.');
        expect(error.code).toEqual(401);
      }
    });

    test('should validate creds from query and send jwt', async () => {
      helpers.validateCredentials.mockImplementationOnce(() => {
        return true;
      });

      jwt.sign.mockImplementationOnce(() => {
        return 'abc';
      });

      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );

      const result = await authPlugin.authenticate({
        query: { username: 'foo', password: 'bar' },
      });

      expect(result.token).toEqual('abc');
      expect(result.expires).toBeGreaterThan(Date.now());
    });

    test('should validate creds from body and send jwt', async () => {
      helpers.validateCredentials.mockImplementationOnce(() => {
        return true;
      });

      jwt.sign.mockImplementationOnce(() => {
        return 'abc';
      });

      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );

      const result = await authPlugin.authenticate({
        body: { username: 'foo', password: 'bar' },
      });

      expect(result.token).toEqual('abc');
      expect(result.expires).toBeGreaterThan(Date.now());
    });
  });

  describe('authorize', () => {
    test('should fail to authorize due to missing token', async () => {
      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );

      try {
        await authPlugin.authorize({ query: {}, headers: [] });
        fail('should have thrown');
      } catch (error) {
        expect(error.message).toEqual('No authorization token.');
        expect(error.code).toEqual(401);
      }
    });

    test('should fail to authorize due to invalid token', async () => {
      jwt.verify.mockImplementationOnce(() => {
        throw new Error('failed to verify token');
      });

      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );

      try {
        await authPlugin.authorize({ query: { token: 'bar' } });
        fail('should have thrown');
      } catch (error) {
        expect(error.message).toEqual('failed to verify token');
        expect(error.code).toEqual(401);
      }
    });

    test('should authorize and send decoded jwt', async () => {
      jwt.verify.mockImplementationOnce(() => {
        return 'abc';
      });

      const expires = Date.now() + 5 * 60 * 1000;
      const token = jwt.sign(
        { exp: Math.floor(expires / 1000), sub: 'fezzik' },
        secret,
      );

      const authPlugin = require('./index.js')(
        secret,
        path.join(__dirname, '../test/fixtures/user-store.json'),
      );

      const result = await authPlugin.authorize({ query: { token } });

      expect(result).toEqual('abc');
    });
  });
});


src/validate-credentials.spec.js
const fs = require('fs-extra');
const { validateCredentials } = require('./validate-credentials');

jest.spyOn(fs, 'readFile');

describe('validate credentials', () => {
  test('should validate', async () => {
    fs.readFile.mockImplementationOnce(() => {
      return Buffer.from(JSON.stringify([
        { username: 'foo', password: 'bar' }
      ]));
    });

    const result = await validateCredentials('foo', 'bar');
    expect(result).toEqual(true);
  });

  test('should invalidate if user not found', async () => {
    fs.readFile.mockImplementationOnce(() => {
      return Buffer.from(JSON.stringify([
        { username: 'fooz', password: 'bar' }
      ]));
    });

    const result = await validateCredentials('foo', 'bar');
    expect(result).toEqual(false);
  });

  test('should invalidate if password does not match', async () => {
    fs.readFile.mockImplementationOnce(() => {
      return Buffer.from(JSON.stringify([
        { username: 'foo', password: 'baz' }
      ]));
    });

    const result = await validateCredentials('foo', 'bar');
    expect(result).toEqual(false);
  });

  test('should throw error reading store', async () => {
    fs.readFile.mockImplementationOnce(() => {
      throw new Error('file error');
    });

    try {
      await validateCredentials('foo', 'bar');
    } catch (err) {
      expect(err.message).toEqual('Auth plugin: error reading auth store');
    }    
  });
});


jest.config.js
/**
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

/** @type {import('jest').Config} */
const config = {
  testTimeout: 20000,

  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/private/var/folders/1p/btf52r0540b8y4pwhjt_k30r6sg9rm/T/jest_3rrqqc",

  // Automatically clear mock calls, instances, contexts and results before every test
  // clearMocks: false,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  // coverageDirectory: undefined,

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: 'v8',

  // A list of reporter names that Jest uses when writing coverage reports
  coverageReporters: ['json', 'text', 'lcov', 'clover'],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // The default configuration for fake timers
  // fakeTimers: {
  //   "enableGlobally": false
  // },

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "mjs",
  //   "cjs",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "json",
  //   "node"
  // ],

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  // preset: undefined,

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state before every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state and implementation before every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // The number of seconds after which a test is considered as slow and reported as such in the results.
  // slowTestThreshold: 5,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  // testEnvironment: "jest-environment-node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  testMatch: ['<rootDir>/src/**/*.spec.[jt]s?(x)'],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jest-circus/runner",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/",
  //   "\\.pnp\\.[^\\/]+$"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};

module.exports = config;


src/index.js
const fs = require('fs');
const jwt = require('jsonwebtoken');
const joi = require('joi');
const { validateCredentials } = require('./validate-credentials');
const TOKEN_EXPIRATION_MINUTES = 60;

const optionsSchema = joi.object({
  tokenExpirationMinutes: joi
    .number()
    .integer()
    .greater(4)
    .default(TOKEN_EXPIRATION_MINUTES),
}).unknown();

let _tokenExpirationMinutes;
let _secret;
let _userStoreFilePath;

/**
 * configure auth functions
 * @param {string}   secret - secret for generating tokens
 * @param {string}   userStoreFilePath - file path of user store JSON file
 * @param {object}   options
 * @param {integer}  options.tokenExpirationMinutes - number of minutes until token expires
 */
function initAuthPlugin(secret, userStoreFilePath, options = {}) {
  // Throw error if user-store file does not exist
  const result = fs.existsSync(userStoreFilePath);
  if (!result) {
    throw new Error(`${userStoreFilePath} not found`);
  }
  _secret = secret;
  _userStoreFilePath = userStoreFilePath;

  const {
    error,
    value: { tokenExpirationMinutes },
  } = optionsSchema.validate(options);

  if (error) {
    throw new Error(error.details[0].message);
  }

  _tokenExpirationMinutes = tokenExpirationMinutes;

  return {
    type: 'auth',
    authenticate,
    authorize,
    name: 'file-based-auth-store',
    version: require('../package.json').version
  };
}


/**
 * Authenticate a user's submitted credentials
 * @param {object} req Express request object
 * @returns {Promise}
 */
async function authenticate(req) {
  const expires = Date.now() + _tokenExpirationMinutes * 60 * 1000;
  const { query = {}, body = {} } = req;
  const { username, password, token } = { ...query, ...body };

  if (token) {
    const { sub } = decodeToken(token);
    return {
      token: createToken(sub, expires),
      expires,
    };
  }

  // Validate user's credentials
  const valid = await validateCredentials(
    username,
    password,
    _userStoreFilePath,
  );

  if (!valid) {
    let err = new Error('Invalid credentials.');
    err.code = 401;
    throw err;
  }

  // Create access token and wrap in response object
  
  return {
    token: createToken(username, expires),
    expires,
  };
}

/**
 * Validate a token
 * @param {object} req Express request object
 * @returns {Promise}
 */
async function authorize(req) {
  const { query = {}, body = {} } = req;
  const headerToken = req.headers['authorization'] ?? req.headers['authorization'].replace(/^Bearer /, '');
  const params = { ...query, ...body };
  const token = headerToken || params.token;

  if (!token) {
    let err = new Error('No authorization token.');
    err.code = 401;
    throw err;
  }
  // Verify token with async decoded function
  return decodeToken(token);
}

async function decodeToken(token) {
  try {
    const decoded = await jwt.verify(token, _secret);
    return decoded;
  } catch (err) {
    err.code = 401;
    throw err;
  }
}

function createToken(sub, expires) {
  return jwt.sign({ exp: Math.floor(expires / 1000), sub }, _secret);
}

module.exports = initAuthPlugin;
